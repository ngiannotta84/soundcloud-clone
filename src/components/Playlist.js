import React from "react";
import PropTypes from "prop-types";
import { music, play } from "../media/icons";
import "../styles/playlist.css";

const Playlist = ({
  playlist,
  playlistIndex,
  setPlaylistIndex,
  removeFromPlaylist,
  setPlaying,
}) => {
  const handleSetPlaying = (i) => {
    setPlaylistIndex(i);
    setPlaying(true);
  };

  return (
    <div className="playlist">
      {playlist.map((song, i) => {
        return (
          <div
            key={song.key}
            className={`playlist__song ${
              playlistIndex === i && "play__song-active"
            }`}
          >
            <div className="playlist__song-left">
              <img
                src={song.image || music}
                alt={`${song.albumName} cover art`}
                className="playlist__song-cover-art"
              />
            </div>
            <div className="playlist__song-right">
              <div className="playlist__song-info">
                <p className="playlist__song-artist">{song.artistName}</p>
                <p className="playlist__song-name">{song.songName}</p>
              </div>
              <button
                type="button"
                className="playlist__play-now"
                onClick={() => handleSetPlaying(i)}
              >
                <img src={play} alt="play now" />
              </button>
              <button
                type="button"
                className="playlist__remove"
                onClick={() => removeFromPlaylist(i)}
              >
                x
              </button>
            </div>
          </div>
        );
      })}
    </div>
  );
};

Playlist.defaultProps = {
  playlist: null,
};

Playlist.propTypes = {
  playlist: PropTypes.arrayOf(
    PropTypes.shape({
      image: PropTypes.string,
      artistName: PropTypes.string,
      albumName: PropTypes.string,
      audio: PropTypes.string,
      songName: PropTypes.string,
    })
  ),
  playlistIndex: PropTypes.number.isRequired,
  setPlaylistIndex: PropTypes.func.isRequired,
  removeFromPlaylist: PropTypes.func.isRequired,
  setPlaying: PropTypes.func.isRequired,
};

export default Playlist;
