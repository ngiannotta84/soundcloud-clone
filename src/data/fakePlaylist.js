import card from "./images/card.png";
import dodl from "./images/Final Render_Cropped 3000.png";
import slumber from "./songs/Slumber.flac";
import careless from "./songs/Your Careless Embrace.flac";
import restless from "./songs/Restless Thoughts.flac";
import detunedLove from "./songs/Detuned Love.flac";
import falling from "./songs/Falling into the Void.flac";

const playlist = [
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: slumber,
    songName: "Slumber",
  },
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: careless,
    songName: "Your Careless Embrace",
  },
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: restless,
    songName: "Restless Thoughts",
  },
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: detunedLove,
    songName: "Detuned Love",
  },
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: falling,
    songName: "Falling into the Void",
  },
];

export default playlist;
