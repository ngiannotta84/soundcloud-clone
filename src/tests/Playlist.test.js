import React from "react";
import { render, screen } from "@testing-library/react";
import Playlist from "../components/Playlist";
import fakePlaylist from "./testData/fakePlaylist";

describe("Playlist", () => {
  const validProps = {
    playlist: fakePlaylist,
    playlistIndex: 1,
    setPlaylistIndex: jest.fn(),
  };

  test("snapshot", () => {
    const { asFragment } = render(<Playlist {...validProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe("tests", () => {
    beforeEach(() => {
      render(<Playlist {...validProps} />);
    });

    test("renders correctly", () => {
      validProps.playlist.forEach((song) => {
        expect(
          screen.getByAltText(`${song.albumName} cover art`)
        ).toHaveAttribute("src", song.image);
        expect(screen.getByText(song.artistName)).toBeInTheDocument();
        expect(screen.getByText(song.songName)).toBeInTheDocument();
      });
      expect(screen.getAllByText(/play now/i)).toHaveLength(
        validProps.playlist.length
      );
    });
  });
});
