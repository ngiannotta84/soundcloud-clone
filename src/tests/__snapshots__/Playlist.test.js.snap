// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist snapshot 1`] = `
<DocumentFragment>
  <div
    class="playlist"
  >
    <div
      class="playlist__song false"
    >
      <div
        class="playlist__song-left"
      >
        <img
          alt="album 1 cover art"
          class="playlist__song-cover-art"
          src="image 1"
        />
      </div>
      <div
        class="playlist__song-right"
      >
        <div
          class="playlist__song-info"
        >
          <p
            class="playlist__song-artist"
          >
            artist 1
          </p>
          <p
            class="playlist__song-name"
          >
            song 1
          </p>
        </div>
        <button
          class="playlist__song-button"
          type="button"
        >
          Play Now
        </button>
      </div>
    </div>
    <div
      class="playlist__song play__song-active"
    >
      <div
        class="playlist__song-left"
      >
        <img
          alt="album 2 cover art"
          class="playlist__song-cover-art"
          src="image 2"
        />
      </div>
      <div
        class="playlist__song-right"
      >
        <div
          class="playlist__song-info"
        >
          <p
            class="playlist__song-artist"
          >
            artist 2
          </p>
          <p
            class="playlist__song-name"
          >
            song 2
          </p>
        </div>
        <button
          class="playlist__song-button"
          type="button"
        >
          Play Now
        </button>
      </div>
    </div>
    <div
      class="playlist__song false"
    >
      <div
        class="playlist__song-left"
      >
        <img
          alt="album 3 cover art"
          class="playlist__song-cover-art"
          src="image 3"
        />
      </div>
      <div
        class="playlist__song-right"
      >
        <div
          class="playlist__song-info"
        >
          <p
            class="playlist__song-artist"
          >
            artist 3
          </p>
          <p
            class="playlist__song-name"
          >
            song 3
          </p>
        </div>
        <button
          class="playlist__song-button"
          type="button"
        >
          Play Now
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
